File: backend/schemas/plan.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime


class PlanResponse(BaseModel):
    _id: str
    name: str
    description: str
    cropTypes: List[str]
    regions: List[str]
    premiumRange: dict
    coverageAmount: float
    benefits: List[str]
    duration: int
    isActive: bool
    createdAt: Optional[str] = None
    updatedAt: Optional[str] = None


class PlanCreateRequest(BaseModel):
    name: str
    description: str
    cropTypes: List[str]
    regions: List[str]
    premiumRange: dict
    coverageAmount: float
    benefits: List[str]
    duration: int = 12
    isActive: bool = True


class PlanUpdateRequest(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    cropTypes: Optional[List[str]] = None
    regions: Optional[List[str]] = None
    premiumRange: Optional[dict] = None
    coverageAmount: Optional[float] = None
    benefits: Optional[List[str]] = None
    duration: Optional[int] = None
    isActive: Optional[bool] = None

File: models/plan.ts
import mongoose, { Schema, Document, Model } from 'mongoose'

export interface IPlan extends Document {
  _id: mongoose.Types.ObjectId
  name: string
  description: string
  cropTypes: string[]
  regions: string[]
  premiumRange: {
    min: number
    max: number
  }
  coverageAmount: number
  benefits: string[]
  terms: string
  duration: number
  isActive: boolean
}

const PlanSchema = new Schema<IPlan>(
  {
    name: {
      type: String,
      required: true,
      trim: true,
    },
    description: {
      type: String,
      required: true,
    },
    cropTypes: [{
      type: String,
      required: true,
    }],
    regions: [{
      type: String,
      required: true,
    }],
    premiumRange: {
      min: { type: Number, required: true },
      max: { type: Number, required: true },
    },
    coverageAmount: {
      type: Number,
      required: true,
    },
    benefits: [{
      type: String,
    }],
    terms: {
      type: String,
      default: '',
    },
    duration: {
      type: Number,
      required: true,
      default: 12,
    },
    isActive: {
      type: Boolean,
      default: true,
    },
  },
  {
    timestamps: true,
  }
)

const Plan: Model<IPlan> = mongoose.models.Plan || mongoose.model<IPlan>('Plan', PlanSchema)

export default Plan
